File: .orkestrel/campaign/docs-parity/instruments/d7/units/process/d7n-process-converge/cells.py
import re, sys

SPLIT = re.compile(r'(?<!\\)\|')

def rows(path):
	out = {}
	lines = open(path, encoding='utf8').read().split('\n')
	i = 0
	table = 0
	while i < len(lines):
		if lines[i].startswith('|') and i + 1 < len(lines) and re.match(r'^\|[\s:-]+\|', lines[i + 1]):
			header = [c.strip() for c in SPLIT.split(lines[i])[1:-1]]
			j = i + 2
			table += 1
			while j < len(lines) and lines[j].startswith('|'):
				cells = [c.strip() for c in SPLIT.split(lines[j])[1:-1]]
				key = (table, cells[0])
				out[key] = dict(zip(header, cells))
				j += 1
			i = j
			continue
		i += 1
	return out

before = rows(sys.argv[1])
after = rows(sys.argv[2])
compared = 0
mismatch = 0
missing = 0
for key, cells in before.items():
	if key not in after:
		missing += 1
		print('ROW MISSING AFTER:', key)
		continue
	for header, value in cells.items():
		if header in ('Summary', 'Behavior'):
			continue
		compared += 1
		other = after[key].get(header)
		if other != value:
			mismatch += 1
			print('CELL CHANGED:', key, header, repr(value), '->', repr(other))
added = [k for k in after if k not in before]
print(f'rows before: {len(before)}, rows after: {len(after)}, non-Summary cells compared: {compared}, mismatched: {mismatch}, rows missing after: {missing}, rows added: {len(added)}')
for k in added:
	print('ROW ADDED:', k)

File: .orkestrel/campaign/docs-parity/instruments/d7/units/queue/d7n-queue-converge/cells.mjs
// Compares every non-Summary cell of every table row between two guide texts, keyed by
// the row's first cell. Splits on a pipe not preceded by a backslash.
import { readFileSync } from 'node:fs'

function readRows(text) {
	const rows = new Map()
	const lines = text.split('\n')
	let header
	let table = 0
	for (const line of lines) {
		if (!line.startsWith('| ')) {
			header = undefined
			continue
		}
		const cells = line
			.slice(1, line.endsWith('|') ? -1 : undefined)
			.split(/(?<!\\)\|/)
			.map((cell) => cell.trim())
		if (cells.every((cell) => /^-+$/.test(cell))) continue
		if (header === undefined) {
			header = cells
			table += 1
			continue
		}
		const compared = new Set(['Summary', 'Behavior'])
		const kept = cells.filter((_, index) => !compared.has(header[index]))
		rows.set(`${table}:${header[0]}:${cells[0]}`, kept.join(' | '))
	}
	return rows
}

const before = readRows(readFileSync(process.argv[2], 'utf8'))
const after = readRows(readFileSync(process.argv[3], 'utf8'))
const changed = []
const missing = []
for (const [key, value] of before) {
	if (!after.has(key)) missing.push(key)
	else if (after.get(key) !== value) changed.push(`${key}\n  before: ${value}\n  after:  ${after.get(key)}`)
}
console.log(`rows before: ${before.size}, rows after: ${after.size}`)
console.log(`rows missing after: ${missing.length}${missing.length > 0 ? `\n  ${missing.join('\n  ')}` : ''}`)
console.log(`non-Summary cells changed: ${changed.length}${changed.length > 0 ? `\n${changed.join('\n')}` : ''}`)
